src/TesApi.Web/BatchScheduler.cs

```diff
@@ -361,14 +363,89 @@ await foreach (var cloudPool in GetCloudPools())
             }
         }
 
+
+        /// <inheritdoc/>
+        public async IAsyncEnumerable<(TesTask TesTask, Task<bool> IsModified)> ProcessTesTasksAsync(IEnumerable<TesTask> tesTasks, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            var (TesTasks, State) = await GetState(tesTasks.ToList());
+
+            await foreach(var item in WhenEach(TesTasks.ToAsyncEnumerable().SelectAwait(async t => await ProcessTesTaskAsync(t, State)).ToEnumerable(), t => t.IsModified, cancellationToken))
+            {
+                yield return item;
+            }
+
+            async Task<(List<TesTask> TesTasks, (IReadOnlyDictionary<CloudPool, IReadOnlyList<ComputeNode>> PoolsAndNodes, IReadOnlyDictionary<CloudJob, IReadOnlyList<CloudTask>> JobsAndTasks) State)> GetState(List<TesTask> listOfTasks)
+                => (listOfTasks, await azureProxy.GetBatchAccountStateAsync(
+                enableBatchAutopool,
+                enableBatchAutopool
+                    ? listOfTasks.Select(t => t.Id)
+                    : GetPools().Select(p => p.Pool.PoolId),
+                cancellationToken));
+        }
+
+        // Adapted from https://stackoverflow.com/a/70026589
+        private static async IAsyncEnumerable<T> WhenEach<T>(IEnumerable<T> items, Func<T, Task> getTask, [EnumeratorCancellation] CancellationToken cancellationToken)
```

Comment:
do we really need this?

Reply:
Simpler means that all the tasks are batched and clumped at each stage of processing, which would only increase API retries due to always hammering the same set of resources at the same time and eliminating much (if not most) of the benefit of this effort.

Doing it this way will mean that task updates flow back to the database as soon as they are determined, which will speed up workflows overall and prevent issues related to one task from delaying reporting of all other tasks.